types/jest/index.d.ts

```diff
@@ -72,6 +72,8 @@ type ExtractEachCallbackArgs<T extends ReadonlyArray<any>> = {
         : 'fallback'
 ];
 
+type MockModuleFactory<T> = () => T extends { default: any } ? Partial<T> & { __esModule: boolean } : Partial<T>;
```

Comment:
```suggestion
type MockModuleFactory<T> = () => T extends { default: any } ? T & { __esModule: true } : T;
```

When you provide a mock jest doesn't apply the mock partially or shallow merges it. So you need to provide all exports. Let me give you an example to make things clear

```ts
// ./sample-module.ts
export const a = "a";
export const b = "b";

// ./setup-mocks.ts
jest.doMock("./sample-module", () => ({ a: "MOCKED-A" }))

// ./stuff.test.ts
import "./setup-mocks"
import { a, b } from "./sample-module"
test("stuff", () => {
	expect(a.toUpperCase()).toBe("MOCKED-A");
	expect(b.toUpperCase()).toBe("B");
	// ^^ no compile time errors in the above statement
	// but in runtime you would get `TypeError: Cannot read property 'toUpperCase' of undefined`
	// so mocks should fulfill the contract that importing sample-module provides that is b will be a string
})

```
A [demo](https://repl.it/@devanshj/DefinitelyTyped-PR-46072) of the above scenario (press the "Run" button)

I also usually don't like violations of `no‑unnecesary‑generics` but in this case it's giving a nice DX especially when coupled with the `__esModule` thing which I like as it's required by jest as you said. Like for example:

```ts
//./some-module
export default () => "baz";

// ./index.spec.ts
doMock<typeof import("./some-module")>(
  "./some-module",
  () => ({})
  // error as expected coz default is missing
) 

doMock<typeof import("./some-module")>(
  "./some-module",
  () => ({ default: () => "baz" })
  // still error as expected because didn't pass __esModule
)
```
A [demo](https://codesandbox.io/s/loving-hill-m236g?file=/src/index.ts) of the above. Also if you can it would be nice if you can make a test like above by making a new file as module an expecting errors.

Now comes why `__esModule: true` instead of `__esModule?: boolean`? (Disclaimer I might be wrong on this) Because when you're working in a es module system you can't export with a key default so people who need `__esModule: false` or `__esModule: undefined` are the ones that use commonjs and have `module.exports.default` and don't use a transpiler or something (tbh I'm not sure when you would have a default export with __esModule not set as true so I'm guessing this would be the case)... so `__esModule: true` type definition is gonna annoy them BUT the tradeoff is the people who work in es module system (which I guess would be the majority by huge margin) maybe are gonna save hours that might have been wasted because the forgot to pass `__esModule: true` so it's evil good.

Also would have been cool as heck if `import` didn't expect a string literal haha but it rightfully does expect it otherwise would have made the definition `<M extends string, T = typeof import(M)>(module: M, factory: () => T extends { default: any } ? T & { __esModule: true } : T ): jest`

Reply:
A better definition would be:
```suggestion
type MockModuleFactory<T> = () => T extends { default: unknown }
	? T & { __esModule: true }
	: T;
```

And the mock function would then use:
```ts
function mock<M extends string>(
	module: M,
	factory: MockModuleFactory<typeof import(M)>,
): typeof jest;
```

But sadly, the `import(…)` type can’t currently support a generic parameter. (see <https://github.com/microsoft/TypeScript/issues/31090>)

Reply:
I probably think `unknown` would be better than `any` but I can't really put a finger on which cases like what difference will it actually make?

Reply:
> When you provide a mock jest doesn't apply the mock partially or shallow merges it. So you need to provide all exports.

Aye, am aware of that 🙂. The idea with the Partial was to allow devs to only provide the minimum they need for a test. 
An example would be testing something which imported a single item from a 3rd-party package. That 3rd-party package could contain numerous exports, but from the test's perspective only the one imported item in use needs to be overridden.

I guess for that kind of scenario, the developer could manually make the generic value `Partial<typeof externalPackage>` or alternatively make use of `jest.requireActual` and merge the originals with specific override.

Following on with the Partial chain of thought, `__esModule: boolean` felt more accurate, as the `default` may not actually be required.

-----

Thank you both for the good suggestions and thank you @devanshj for the detailed reply 🙂.

I have updated the PR with the suggested changes (no partial, use `unknown`, no overload and enforced `__esModule: true`) and added additional tests to check the negatives (errors correctly occur).

Reply:
> Aye. Am aware of that 🙂. 

Oh haha cool

> I guess for that kind of scenario, the developer could manually make the generic value Partial<typeof externalPackage>

Yep that would be better

> Thank you both for the good suggestions and thank you @devanshj for the detailed reply 🙂

You're welcome! 😁